x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java

```diff
@@ -133,9 +137,69 @@ private static Decision debugYesAllowed(RoutingAllocation allocation, List<Strin
      * Given a string of comma-separated prioritized tiers (highest priority
      * first) and an allocation, find the highest priority tier for which nodes
      * exist. If no nodes for any of the tiers are available, returns an empty
-     * {@code Optional<String>}.
+     * {@code Optional<String>}. This method takes into account the desired nodes
+     * in order to know if there are planned topology changes in the cluster
+     * that can remove a tier that's part of the cluster now.
      */
-    public static Optional<String> preferredAvailableTier(List<String> prioritizedTiers, DiscoveryNodes nodes) {
+    public static Optional<String> preferredAvailableTier(List<String> prioritizedTiers, DiscoveryNodes nodes, DesiredNodes desiredNodes) {
+        final var desiredNodesPreferredTier = getPreferredTierFromDesiredNodes(prioritizedTiers, nodes, desiredNodes);
+
+        if (desiredNodesPreferredTier.isPresent()) {
+            return desiredNodesPreferredTier;
+        }
+
+        return getPreferredAvailableTierFromClusterMembers(prioritizedTiers, nodes);
+    }
+
+    /**
+     * Given a list of tiers in descending order, return the tier that's present
+     * in the desired nodes with the highest priority, if none is present returns an
+     * {@code Optional.empty()}.
+     */
+    public static Optional<String> getPreferredTierFromDesiredNodes(
+        List<String> prioritizedTiers,
+        DiscoveryNodes discoveryNodes,
+        DesiredNodes desiredNodes
+    ) {
+        if (desiredNodes == null) {
+            return Optional.empty();
+        }
+
+        for (int i = 0; i < prioritizedTiers.size(); i++) {
+            final var tier = prioritizedTiers.get(i);
+            final var nextTier = i + 1 == prioritizedTiers.size() ? null : prioritizedTiers.get(i + 1);
```

Comment:
I think we should loop through all possible lower tiers? Suppose we have cold,warm,hot preference but warm is unused. Then we would want to do the check here against hot too?